File: analyzer-direct/src/main/scala/io/github/tassiLuca/analyzer/lib/Analyzer.scala
package io.github.tassiLuca.analyzer.lib

import gears.async.{Async, AsyncOperations}
import io.github.tassiLuca.analyzer.commons.lib
import io.github.tassiLuca.analyzer.commons.lib.RepositoryReport
import io.github.tassiLuca.dse.boundaries.CanFail

/** A generic analyzer of organization/group/workspace repositories. */
trait Analyzer:

  /** Performs a **suspending** analysis of the [[organizationName]]'s repositories, providing the results
    * incrementally to the [[updateResults]] function. It may fail along the way!
    * @return the overall results of the analysis.
    */
  def analyze(organizationName: String)(
      updateResults: RepositoryReport => Unit,
  )(using Async, AsyncOperations, CanFail): Seq[RepositoryReport]

object Analyzer:
  /** @return the basic version of the [[Analyzer]], i.e. the one performing suspending
    *         operation until all repositories are fetched.
    */
  def basic(service: RepositoryService): Analyzer = BasicAnalyzer(service)

  /** @return an incremental version of the [[Analyzer]] which reacts to each fetched repositories
    *         making the computation faster when dealing with a high number of repositories.
    */
  def incremental(service: RepositoryService): Analyzer = IncrementalAnalyzer(service)

  /** @return a version of the [[Analyzer]] internally using the new Flow abstraction. */
  def flowing(service: RepositoryService): Analyzer = FlowingAnalyzer(service)

File: analyzer-direct-kt/src/main/kotlin/io/github/tassiLuca/analyzer/lib/Analyzer.kt
package io.github.tassiLuca.analyzer.lib

/** A generic analyzer of organization/group/workspace repositories. */
interface Analyzer {

    /**
     * Performs a **suspending** analysis of the [organizationName]'s repositories, providing the results
     * incrementally to the [updateResults] function.
     * @return a successful [Result], containing the set of [RepositoryReport]s or a failure [Result] if the analysis.
     */
    suspend fun analyze(
        organizationName: String,
        updateResults: suspend (RepositoryReport) -> Unit = { _ -> },
    ): Result<Set<RepositoryReport>>

    companion object {
        /** Creates a new GitHub organization [Analyzer] based on Coroutines `Channel`s. */
        fun ofGitHubByChannels(gitHubProvider: GitHubRepositoryProvider): Analyzer =
            GitHubAnalyzerByChannels(gitHubProvider)

        /** Creates a new GitHub organization [Analyzer] based on Coroutines `Flow`s. */
        fun ofGithubByFlows(gitHubProvider: GitHubRepositoryProvider): Analyzer =
            GitHubAnalyzerByFlows(gitHubProvider)
    }
}
